### ISSUE TITLE
calwebb_image3.tweakreg to better handle runaway processing

### ISSUE THREAD

**Issue by stscijgbot** [spacetelescope/jwst#3009]
Issue [JP-498](https://jira.stsci.edu/browse/JP-498) was created by Jonathan Eisenhamer:

# Issue

During SDP runs, proposal jw00752, observation 8, a MIRI mosaic, while being processed by `calwebb_image3`, was running for over a week. Previous runs had noted 24+ hour processing before being manually killed off.

The data was provided by the OTB and is known to be bad. However, it was been requested that the pipeline be able to detect and handle bad data/runaway processing.

**Comment by stscieisenhamer** [comment]
# Resources

Data is located in
```
/grp/jwst/ssb/eisenhamer/jwst_issues/issue3009
```
<details><summary>Log of the hanging run...</summary>
<p>

```
$ strun cfgs/calwebb_image3.cfg jw00752-o008_20181219t183355_image3_001_asn.json
2019-01-04 08:49:22,120 - stpipe.Image3Pipeline - INFO - Image3Pipeline instance created.
2019-01-04 08:49:22,122 - stpipe.Image3Pipeline.tweakreg - INFO - TweakRegStep instance created.
2019-01-04 08:49:22,123 - stpipe.Image3Pipeline.skymatch - INFO - SkyMatchStep instance created.
2019-01-04 08:49:22,125 - stpipe.Image3Pipeline.outlier_detection - INFO - OutlierDetectionStep instance created.
2019-01-04 08:49:22,126 - stpipe.Image3Pipeline.resample - INFO - ResampleStep instance created.
2019-01-04 08:49:22,127 - stpipe.Image3Pipeline.source_catalog - INFO - SourceCatalogStep instance created.
2019-01-04 08:49:22,127 - stpipe - INFO - Hostname: inanna2.local.stsci.edu
2019-01-04 08:49:22,127 - stpipe - INFO - OS: Darwin
2019-01-04 08:49:22,161 - stpipe.Image3Pipeline - INFO - Step Image3Pipeline running with args ('jw00752-o008_20181219t183355_image3_001_asn.json',).
2019-01-04 08:49:23,090 - stpipe.Image3Pipeline - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/asdf/asdf.py:202: UserWarning: File was created with extension 'asdf.extension.BuiltinExtension' from package asdf-2.4.0.dev1697, but older version asdf-2.4.0.dev1695 is installed
  warnings.warn(fmt_msg)

2019-01-04 08:49:23,090 - stpipe.Image3Pipeline - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/asdf/asdf.py:202: UserWarning: File was created with extension 'astropy.io.misc.asdf.extension.AstropyAsdfExtension' from package astropy-3.2.dev23666, but older version astropy-3.2.dev23658 is installed
  warnings.warn(fmt_msg)

2019-01-04 08:49:23,091 - stpipe.Image3Pipeline - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/asdf/asdf.py:202: UserWarning: File was created with extension 'astropy.io.misc.asdf.extension.AstropyExtension' from package astropy-3.2.dev23666, but older version astropy-3.2.dev23658 is installed
  warnings.warn(fmt_msg)

2019-01-04 08:49:23,091 - stpipe.Image3Pipeline - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/asdf/asdf.py:202: UserWarning: File was created with extension 'jwst.transforms.jwextension.JWSTExtension' from package jwst-0.13.0b0.dev157+g99e1588a, but older version jwst-0.13.0b0.dev33 is installed
  warnings.warn(fmt_msg)

2019-01-04 08:49:23,892 - stpipe.Image3Pipeline - INFO - Prefetching reference files for dataset: 'jw00752008001_02105_00001_mirimage_cal.fits' reftypes = ['drizpars']
2019-01-04 08:49:28,275 - stpipe.Image3Pipeline - INFO - Prefetch for DRIZPARS reference file is '/Users/eisenham/Downloads/crds-cache/references/jwst/miri/jwst_miri_drizpars_0001.fits'.
2019-01-04 08:49:29,957 - stpipe.Image3Pipeline - INFO - Prefetching reference files for dataset: 'jw00752008001_02103_00001_mirimage_cal.fits' reftypes = ['drizpars']
2019-01-04 08:49:29,960 - stpipe.Image3Pipeline - INFO - Prefetch for DRIZPARS reference file is '/Users/eisenham/Downloads/crds-cache/references/jwst/miri/jwst_miri_drizpars_0001.fits'.
2019-01-04 08:49:32,018 - stpipe.Image3Pipeline - INFO - Prefetching reference files for dataset: 'jw00752008001_02107_00001_mirimage_cal.fits' reftypes = ['drizpars']
2019-01-04 08:49:32,021 - stpipe.Image3Pipeline - INFO - Prefetch for DRIZPARS reference file is '/Users/eisenham/Downloads/crds-cache/references/jwst/miri/jwst_miri_drizpars_0001.fits'.
2019-01-04 08:49:36,027 - stpipe.Image3Pipeline - INFO - Prefetching reference files for dataset: 'jw00752008001_02101_00001_mirimage_cal.fits' reftypes = ['drizpars']
2019-01-04 08:49:36,030 - stpipe.Image3Pipeline - INFO - Prefetch for DRIZPARS reference file is '/Users/eisenham/Downloads/crds-cache/references/jwst/miri/jwst_miri_drizpars_0001.fits'.
2019-01-04 08:49:36,030 - stpipe.Image3Pipeline - INFO - Starting calwebb_image3 ...
2019-01-04 08:49:37,587 - stpipe.Image3Pipeline - INFO - Aligning input images...
2019-01-04 08:49:37,626 - stpipe.Image3Pipeline.tweakreg - INFO - Step tweakreg running with args (<ModelContainer>,).
2019-01-04 08:49:38,744 - stpipe.Image3Pipeline.tweakreg - INFO - Detected 1 sources in jw00752008001_02105_00001_mirimage_cal.fits.
2019-01-04 08:49:39,533 - stpipe.Image3Pipeline.tweakreg - INFO - Detected 1 sources in jw00752008001_02103_00001_mirimage_cal.fits.
2019-01-04 08:49:40,411 - stpipe.Image3Pipeline.tweakreg - INFO - Detected 1 sources in jw00752008001_02107_00001_mirimage_cal.fits.
2019-01-04 08:49:41,283 - stpipe.Image3Pipeline.tweakreg - INFO - Detected 1 sources in jw00752008001_02101_00001_mirimage_cal.fits.
2019-01-04 08:49:41,360 - stpipe.Image3Pipeline.tweakreg - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/spherical_geometry/polygon.py:590: RuntimeWarning: invalid value encountered in triple_product
  orient = great_circle_arc.triple_product(A-B, C-B, B)

2019-01-04 08:49:41,361 - stpipe.Image3Pipeline.tweakreg - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/spherical_geometry/polygon.py:133: RuntimeWarning: invalid value encountered in triple_product
  orient = great_circle_arc.triple_product(A-B, C-B, B)

2019-01-04 08:49:41,411 - stpipe.Image3Pipeline.tweakreg - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/spherical_geometry/vector.py:153: RuntimeWarning: invalid value encountered in normalize
  math_util.normalize(xyz, output)

2019-01-04 08:49:41,412 - stpipe.Image3Pipeline.tweakreg - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/spherical_geometry/graph.py:240: RuntimeWarning: invalid value encountered in less
  diff = np.all(np.abs(point - node_array) < 2 ** -32, axis=-1)

2019-01-04 08:49:41,413 - stpipe.Image3Pipeline.tweakreg - WARNING - /Users/eisenham/anaconda3/envs/jwstdev/lib/python3.7/site-packages/spherical_geometry/great_circle_arc.py:284: RuntimeWarning: invalid value encountered in intersects_point
  return math_util.intersects_point(A, B, C)
```

</p>
</summary>

**Comment by jdavies-st** [comment]
@mcara, any ideas what could be going on here?  Clearly there are no sources, so `tweakreg` should be skipped.  That said, the data looks crazy - nothing like what is expected from that detector.

<img width="1029" alt="screen shot 2019-01-04 at 10 42 16 am" src="https://user-images.githubusercontent.com/17088869/50696428-711f3280-100d-11e9-96be-f961278070e3.png">

**Comment by jdavies-st** [comment]
Also, per the log, only a single source is detected in each.  Clearly there are none.  Doesn't there need to be a minimum of 3 sources required for `tweakreg` to proceed?

**Comment by stscieisenhamer** [comment]
About the data: It is from the OTB and, currently, any data out of the OTB is not expected to make sense. In general DMS discussion however, it is felt that the pipelines should be more resilient in such situations. Even when the data is "live", there is no guarantee of "reasonable data".

Regardless, a fair statement can be made that even for the average user, having sanity checks to prevent runaway processing would be a really nice thing.

**Comment by mcara** [comment]
GIGO is almost a postulate. However, if pipeline is going to detect "GI", it should be done, IMO, at the earliest possible stages, long before tweakreg is called. Maybe make an independent step to detect bad data and skip/abort the rest of the processing.

**Comment by jdavies-st** [comment]
`tweakreg` is the only step in the image3 pipeline that cares about GI, as it has the need of real sources at real pointings.  

Resample doesn't care about sources, only about pointings overlapping.  Same with background matching.

**Comment by jdavies-st** [comment]
A possible solution is to just take those warnings from `spherical_geometry` seriously.  Convert them to errors and handle them.

**Comment by stscieisenhamer** [comment]
If this current situation is "just a bug" and the runaway processing is resolved by fixing "the bug", then nothing more need be done then bug a'fixn.

However, if the runaway processing is a valid consequence with the given data, then a reasonable expectation is that the step involved recognizes this situation and reports as such, most likely as an exception.

**Comment by mcara** [comment]
@stscieisenhamer I am curious: is this a new dataset or did the test started failing recently (if so, when)?

**Comment by jdavies-st** [comment]
The issue here is that all four input images have no sources, and are, in fact, the same image with the exact same pixel values and header `wcsinfo`.  💩 

It's serendipity that source is actually detected, and of course it is detected in all four images in the exact same location.

Basically, it probably is a bug in `spherical_geometry`, and perhaps those `RuntimeWarning` should be `RuntimeError` so it doesn't not get into this infinite loop state.  That said, we can handle them in `tweakreg` so it never gets to that point.  That's what we'll do for now, and maybe file an issue in `spherical_geometry` once a more simple test case can be constructed.

**Comment by mcara** [comment]
Also, what is _"proposal jw00752, observation 8"_. Is this for real?

**Comment by mcara** [comment]
The issue actually is that `gWCS` returns `nan` when the coordinate is 0.5 pixels outside the image. I intentionally enlarge the rectangle as to include the outer "edge" of the pixels (not only their centers). It seems that `gWCS` does not like this.

**Comment by nden** [comment]
He, it's fashionable to blame gwcs but it doesn't care that much about the bounding box, which simply tells you where the transform is valid. For MIRI images it happens to be smaller than the image size. If it makes sense for your application, you can disable it when calling the WCS by passing the flag `with_bounding_box=False`. OTH, if you want to transform all pixels, you can get the valid pixels by
```
from gwcs import wcstools
x, y = wcstools.grid_from_bounding_box(wcs.bounding_box)
```

**Comment by mcara** [comment]
_"it's fashionable to blame gwcs"_ sounds like it is blamed undeservingly :)

**Comment by mcara** [comment]
For the record, tweakreg was implemented in the pipeline before `bounding_box` was even available  so it could not use it at that time.

**Comment by mcara** [comment]
@stscieisenhamer I have no idea of what has happened but it appears that somehow I closed this issue. It should be solved via #3015. Feel free to re-open it if necessary.
